pkg/pillar/cmd/zedrouter/appnumonunet.go

```diff
@@ -0,0 +1,303 @@
+		log.Errorf("uNetAppNumAlloc(%s-%s): app(%s) fail: %s\n",
+			netInstStatus.DisplayName, netInstStatus.UUID.String(),
+			appID, errStr)
+		return types.AppNumInvalid, errors.New(errStr)
+	}
+	return appNum, nil
+}
+
+func uNetAppNumDelete(ctx *zedrouterContext,
+	netInstStatus *types.NetworkInstanceStatus, appID string, appNum int) {
+	bitMapAppNumDelete(ctx, netInstStatus, appNum)
+}
+
+// App Num BitMap Routines
+func bitMapAppNumAdd(ctx *zedrouterContext,
+	netInstStatus *types.NetworkInstanceStatus, appNum int) {
+	if appNum >= types.AppNumStart && netInstStatus != nil &&
+		appNum <= netInstStatus.DhcpAppNumCount &&
+		!netInstStatus.AppNumBitMap.IsSet(appNum) {
+		netInstStatus.AppNumBitMap.Set(appNum)
+	}
+}
+
+func bitMapAppNumDelete(ctx *zedrouterContext,
+	netInstStatus *types.NetworkInstanceStatus, appNum int) {
+	if appNum >= types.AppNumStart && netInstStatus != nil &&
+		appNum <= netInstStatus.DhcpAppNumCount &&
+		netInstStatus.AppNumBitMap.IsSet(appNum) {
+		netInstStatus.AppNumBitMap.Clear(appNum)
+	}
+}
+
+// Persistent Address Map routines
+func persistentAddrMapGet(ctx *zedrouterContext, networkID string,
+	appID string) *types.NetworkInstanceAppNumMap {
+	pub := ctx.pubNetworkInstanceAppNumMap
+	key := persistentAddrMapKey(networkID, appID)
+	if addrMap, _ := pub.Get(key); addrMap != nil {
+		appNumMap := addrMap.(types.NetworkInstanceAppNumMap)
+		return &appNumMap
+	}
+	return nil
+}
+
+func persistentAddrMapAdd(ctx *zedrouterContext,
+	networkID string, appID string, appNumOnUNet int) {
+	pub := ctx.pubNetworkInstanceAppNumMap
+	addrMap := types.NetworkInstanceAppNumMap{
+		NetworkID:    networkID,
+		AppID:        appID,
+		AppNumOnUNet: appNumOnUNet,
+	}
+	pub.Publish(addrMap.Key(), addrMap)
+	return
+}
+
+func persistentAddrMapDelete(ctx *zedrouterContext,
+	networkID string, appID string) {
+	pub := ctx.pubNetworkInstanceAppNumMap
+	key := persistentAddrMapKey(networkID, appID)
+	pub.Unpublish(key)
+	return
+}
+
+// Persistent IPAddress Mapping Restore for Network Instance
+func persistentAddrMapRestoreForNetworkInstance(ctx *zedrouterContext,
+	netInstStatus *types.NetworkInstanceStatus) {
+	networkID := netInstStatus.UUIDandVersion.UUID.String()
+	items := ctx.pubNetworkInstanceAppNumMap.GetAll()
+	for _, item := range items {
+		addrMap := item.(types.NetworkInstanceAppNumMap)
+		if addrMap.NetworkID == networkID {
+			bitMapAppNumAdd(ctx, netInstStatus, addrMap.AppNumOnUNet)
+		}
+	}
+}
+
+// Persistent IPAddress Mapping Delete for Network Instance
+func persistentAddrMapDeleteForNetworkInstance(ctx *zedrouterContext,
+	netInstStatus *types.NetworkInstanceStatus) {
+	if netInstStatus == nil {
+		return
+	}
+	pub := ctx.pubNetworkInstanceAppNumMap
+	items := pub.GetAll()
+	networkID := netInstStatus.UUIDandVersion.UUID.String()
+	for _, item := range items {
+		addrMap := item.(types.NetworkInstanceAppNumMap)
+		if addrMap.NetworkID == networkID {
+			pub.Unpublish(addrMap.Key())
```

Comment:
I'm quite concerned that the bitmap isn't maintained correctly. Do you have assertions (in the form of log.Fatal) that the bifmap is in synch with the allocations?
If not I assume the bitmap handling has bugs. That code must be bullet proof and you need to structure the code so that it is easy for a reader to convince themself that it doesn't have bugs. That calls for encapsulation of all handling of the bitmap.